pyphare/tests/test_pharesee/test_geometry.py

```diff
@@ -0,0 +1,583 @@
+    coarse_particles.add(lower_cell_particles.shift_icell(domain_box.size()))
+
+    boxes = {}
+    for ilvl, boxes_data in refinement_boxes.items():
+
+        level_number = int(ilvl.strip("L")) + 1
+
+        if level_number not in boxes:
+            boxes[level_number] = []
+
+        for boxname, lower_upper in boxes_data.items():
+            refinement_box = Box(lower_upper[0][0], lower_upper[1][0])
+            refined_box = boxm.refine(refinement_box, refinement_ratio)
+            boxes[level_number].append(refined_box)
+
+    # coarse level boxes are arbitrarily divided in 2 patches in the middle
+    middle_cell = round(domain_box.upper / 2)
+    lower_box = Box(0, middle_cell)
+    upper_box = Box(middle_cell + 1, domain_box.upper)
+    boxes[0] = [lower_box, upper_box]
+
+    patch_datas = {}
+
+    for ilvl, lvl_box in boxes.items():
+
+        lvl_cell_width = cell_width / (refinement_ratio ** ilvl)
+
+        if ilvl == 0:
+            lvl_particles = coarse_particles
+        else:
+            level_domain_box = boxm.refine(domain_box, refinement_ratio)
+            lvl_ghost_domain_box = boxm.grow(level_domain_box, domain_layout.particleGhostNbr(interp_order))
+            lvl_particles = Particles(box = lvl_ghost_domain_box)
+
+        if ilvl not in patch_datas:
+            patch_datas[ilvl] = []
+
+        for box in lvl_box:
+
+            ghost_box = boxm.grow(box, 5)
+            origin = box.lower * lvl_cell_width
+            layout = GridLayout(box, origin, lvl_cell_width, interp_order)
+
+            datas = {"Bx": bx(ghost_box, lvl_cell_width, domain_size, origin),
+                     "By": by(ghost_box, lvl_cell_width, domain_size, origin),
+                     "Bz": bz(ghost_box, lvl_cell_width, domain_size, origin),
+                     "Ex": ex(ghost_box, lvl_cell_width, domain_size, origin),
+                     "Ey": ey(ghost_box, lvl_cell_width, domain_size, origin),
+                     "Ez": ez(ghost_box, lvl_cell_width, domain_size, origin),
+                     "particles": lvl_particles.select(ghost_box)
+                     }
+
+            boxed_patch_datas = {}
+            for qty_name, data in datas.items():
+                if qty_name == 'particles':
+                    pdata = ParticleData(layout, data)
+                else:
+                    pdata = FieldData(layout, qty_name, data)
+
+                boxed_patch_datas[qty_name] = pdata
+
+            patch_datas[ilvl].append(boxed_patch_datas)
+
+    patches = {}
+    for ilvl, lvl_patch_datas in patch_datas.items():
+
+        if ilvl not in patches:
+            patches[ilvl] = []
+
+        for patch_datas in lvl_patch_datas:
+            patches[ilvl].append(Patch(patch_datas))
+
+    patch_levels = {}
+    for ilvl, lvl_patches in patches.items():
+        patch_levels[ilvl] = PatchLevel(ilvl, lvl_patches)
+
+    sorted_levels_numbers = sorted(patch_levels)
```

Comment:
some of this function should work regardless of dim, but sorting is something we should note is not one of these things

Reply:
I don't understand what you mean in that context. here we sort by level number, what does it have to do with dim ?

Reply:
yes, my mistake. It's a general comment about sorting list of patches more so.